File: platforms/hacker-rank/M--40--two-pluses/solutions/solution.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
	"strings"
)

/*
 * Complete the 'twoPluses' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts STRING_ARRAY grid as parameter.
 */

type plus struct {
	area  int32
	cells map[int]interface{}
}

func minarr(arr []int) int {
	v := arr[0]
	for _, e := range arr[1:] {
		v = min(v, e)
	}
	return v
}

func plusoverlap(p1, p2 plus) bool {
	for k := range p1.cells {
		if _, ok := p2.cells[k]; ok {
			return true
		}
	}
	return false
}

func twoPluses(grid []string) int32 {
	m, n := len(grid), len(grid[0])

	u, d, l, r := make([][]int, m), make([][]int, m), make([][]int, m), make([][]int, m)
	for i := 0; i < m; i++ {
		u[i] = make([]int, n)
		d[i] = make([]int, n)
		l[i] = make([]int, n)
		r[i] = make([]int, n)
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] != 'G' {
				continue
			}

			u[i][j] = 1
			l[i][j] = 1

			if i > 0 {
				u[i][j] += u[i-1][j]
			}

			if j > 0 {
				l[i][j] += l[i][j-1]
			}
		}
	}

	for i := m - 1; i >= 0; i-- {
		for j := n - 1; j >= 0; j-- {
			if grid[i][j] != 'G' {
				continue
			}

			d[i][j] = 1
			r[i][j] = 1

			if i+1 < m {
				d[i][j] += d[i+1][j]
			}

			if j+1 < n {
				r[i][j] += r[i][j+1]
			}
		}
	}

	maskidx := func(i, j int) int {
		return i*n + j
	}

	pluses := make([]plus, 0)

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] != 'G' {
				continue
			}

			maxarm := minarr([]int{
				u[i][j],
				d[i][j],
				l[i][j],
				r[i][j],
			}) - 1

			for k := 0; k <= maxarm; k++ {
				mask := make(map[int]interface{})

				mask[maskidx(i, j)] = nil
				for d_ := 0; d_ <= k; d_++ {
					mask[maskidx(i-d_, j)] = nil
					mask[maskidx(i+d_, j)] = nil
					mask[maskidx(i, j-d_)] = nil
					mask[maskidx(i, j+d_)] = nil
				}

				area := int32(1 + 4*k)
				pluses = append(pluses, plus{cells: mask, area: area})
			}
		}
	}

	sort.Slice(pluses, func(i, j int) bool {
		return pluses[i].area > pluses[j].area
	})

	maxProd := int32(0)

	for i := 0; i < len(pluses)-1; i++ {
		if pluses[i].area*pluses[i+1].area <= maxProd {
			break
		}
		for j := i + 1; j < len(pluses); j++ {
			prod := pluses[i].area * pluses[j].area
			if prod <= maxProd {
				break
			}

			if !plusoverlap(pluses[i], pluses[j]) {
				maxProd = prod
			}
		}
	}

	return maxProd
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16*1024*1024)

	firstMultipleInput := strings.Split(strings.TrimSpace(readLine(reader)), " ")

	nTemp, err := strconv.ParseInt(firstMultipleInput[0], 10, 64)
	checkError(err)
	n := int32(nTemp)

	// mTemp, err := strconv.ParseInt(firstMultipleInput[1], 10, 64)
	checkError(err)
	// m := int32(mTemp)

	var grid []string

	for i := 0; i < int(n); i++ {
		gridItem := readLine(reader)
		grid = append(grid, gridItem)
	}

	result := twoPluses(grid)

	fmt.Fprintf(writer, "%d\n", result)

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}

File: platforms/hacker-rank/M--40--two-pluses/solutions/solution.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'twoPluses' function below.
#
# The function is expected to return an INTEGER.
# The function accepts STRING_ARRAY grid as parameter.
#


def twoPluses(grid):
    m, n = len(grid), len(grid[0])

    u, d, l, r = [], [], [], []
    for i in range(m):
        u.append([0] * n)  # v: u to d
        d.append([0] * n)  # v: d to u
        l.append([0] * n)  # h: l to r
        r.append([0] * n)  # h: r to l

    for i in range(m):
        for j in range(n):
            if grid[i][j] == "G":
                u[i][j] = 1
                l[i][j] = 1

                if i > 0:
                    u[i][j] += u[i - 1][j]

                if j > 0:
                    l[i][j] += l[i][j - 1]

    for i in range(m - 1, -1, -1):
        for j in range(n - 1, -1, -1):
            if grid[i][j] == "G":
                d[i][j] = 1
                r[i][j] = 1

                if i + 1 < m:
                    d[i][j] += d[i + 1][j]

                if j + 1 < n:
                    r[i][j] += r[i][j + 1]

    def cidx(i, j):
        return i * n + j

    pluses = []
    for i in range(m):
        for j in range(n):
            if grid[i][j] != "G":
                continue

            maxarm = min(u[i][j], d[i][j], l[i][j], r[i][j]) - 1

            for k in range(maxarm + 1):
                mask = 0
                mask |= 1 << cidx(i, j)

                for d_ in range(k + 1):
                    mask |= 1 << cidx(i - d_, j)
                    mask |= 1 << cidx(i + d_, j)
                    mask |= 1 << cidx(i, j - d_)
                    mask |= 1 << cidx(i, j + d_)

                area = 1 + 4 * k
                pluses.append((area, mask))

    pluses.sort(reverse=True)

    maxprod = 0

    for i in range(len(pluses) - 1):
        for j in range(i + 1, len(pluses)):
            prod = pluses[i][0] * pluses[j][0]
            if prod <= maxprod:
                break

            if pluses[i][1] & pluses[j][1] == 0:
                maxprod = prod

    return maxprod


if __name__ == "__main__":
    fptr = open(os.environ["OUTPUT_PATH"], "w")
    first_multiple_input = input().rstrip().split()
    n = int(first_multiple_input[0])
    m = int(first_multiple_input[1])
    grid = []

    for _ in range(n):
        grid_item = input()
        grid.append(grid_item)

    result = twoPluses(grid)
    fptr.write(str(result) + "\n")
    fptr.close()
